### DISCUSSION TITLE
Manhattan connector

### DISCUSSION THREAD

**Discussion by mayorovad** [maxGraph/maxGraph#116] [Development]
Need advice!

I have ported Manhattan connector from https://github.com/mwangm/mxgraph-manhattan-connector mentioned in https://github.com/maxGraph/maxGraph/issues/3 to maxGraph and have some questions:

 - [ ] Licensing - original code is under MIT License, what is correct to way to introduce ported code in terms of licensing?
 - [ ] Publishing - how to correctly publish this code? Only solution I see now is to create PR to this repo, but maybe there are different solutions?

**Comment by tbouffard** [comment]
hi @mayorovad 
About the licensing, if we want to integrate it into maxGraph core, the code must be licensed under the Apache 2.0 license. So you can ask the original author if the code can be relicensed.

As an alternative, we could host the ported code in a dedicated repository hosted in the maxGraph GitHub organization and keep the MIT license. The manhattan connector won't be part of maxGraph core but projects that want to use it could be easily integrate it as an additional dependency.

**Reply by tbouffard** [reply]
@mayorovad do you still want to contribute such connector?
I can create the repository in the maxGraph organization and give you administrator permissions. Then you will be able to push the code.
Alternatively, if you already have a GitHub repository containing the ported code, we can transfer it to the maxGraph organization.

Let me know what you think is best for you.

**Reply by mayorovad** [reply]
@tbouffard yes, I want to contribute, but have no understanding how to organize published code in way it will be available in maxGraph.
Currently, there are large chunk of code in `src/view/style/EdgeStyle` with connector method and different static values. But to make it available for use, I added references to it here
https://github.com/maxGraph/maxGraph/blob/7e9bc662bef51fe543b3628e4ab6ecc0e75781b0/packages/core/src/view/style/StyleRegistry.ts#L61-L69
and here
https://github.com/maxGraph/maxGraph/blob/7e9bc662bef51fe543b3628e4ab6ecc0e75781b0/packages/core/src/util/Constants.ts#L589-L600

Also, there is story for Storybook with connector usage.

**Reply by tbouffard** [reply]
Sorry for the late answer.

From my understanding, here is how this will work if we provide the Manhattan Connector as an external component.
Applications that want to use the Connector will use the `maxGraph` configuration. No need to modify `maxGraph` for that.
The external component will be published on npmjs to be consumed by the applications.

To use the connector in a given application
- add the connector in the dependencies of the application
- in the application code
  - register the connector to make it available to style
  - apply the connector to the edge style

To register the connector, choose a string key to identify it. The key will then be used to reference the connector in style configurations. In the example, I choose `manhattanEdgeStyle` for consistency with the keys used in maxGraph, but this can be whatever key as far as it is unique.
In `maxGraph`, the key are currently listed available in the EDGESTYLE enum, this is to guide users. 
```typescript
// ManhattanConnector is the name of the class provided by the 
StyleRegistry.putValue('manhattanEdgeStyle', ManhattanConnector); 
```

To apply the connector to all edges for instance, use the following:
```typescript
const style = graph.getStylesheet().getDefaultEdgeStyle(); // CellStateStyle
style.edgeStyle = 'manhattanEdgeStyle';
```
As for all `CellStateStyle` configuration, this can also be done while inserting the edge or applied to a dedicated `CellStateStyle` (not the default one)


About the Storybook example, it cannot stay in the maxGraph code base.
The repository holding the Connector code would provide the usage example. The example will use the Connector in the same way as any application.

Tell me if I am not clear and feel free to ask for more details.
I can help you on that: this is a good example of an extension living outside of the maxGraph codebase. And I am pretty sure this will help a lot of people to see it in action.

**Reply by tbouffard** [reply]
@mayorovad hi, do you still want to work on this topic?
If not, could you share the repository where you put the manhattan connector implementation within maxgraph@core? this will let others later extract the code to a dedicated repo.

**Comment by tbouffard** [comment]
@mayorovad Are you still interested in sharing your implementation?

**Reply by mayorovad** [reply]
Hello @tbouffard! Yes, I am! Created PR #232, please, take a look 👀
